File: src/test/java/com/jacuum/leaderboard/JsonLeaderboardTest.java
package com.jacuum.leaderboard;

import com.jacuum.algo.Direction;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import java.nio.file.Path;
import java.time.Instant;
import java.util.List;
import static org.assertj.core.api.Assertions.*;

class JsonLeaderboardTest {

    @TempDir Path tmp;

    private LeaderboardEntry entry() {
        return new LeaderboardEntry(
            "id1", "Luke", "🤖", "hash1", "SMALL", "Random",
            42, 100, 4200, Instant.now().toString(),
            List.of(new TraceEvent(1, Direction.EAST, 1, 1, 100))
        );
    }

    @Test void saveAndRetrieveEntry() throws Exception {
        Path file = tmp.resolve("lb.json");
        Leaderboard lb = new JsonLeaderboard(file);
        lb.save(entry());
        List<LeaderboardEntry> entries = lb.entries();
        assertThat(entries).hasSize(1);
        assertThat(entries.get(0).username()).isEqualTo("Luke");
    }

    @Test void persistsAcrossInstances() throws Exception {
        Path file = tmp.resolve("lb2.json");
        new JsonLeaderboard(file).save(entry());
        List<LeaderboardEntry> entries = new JsonLeaderboard(file).entries();
        assertThat(entries).hasSize(1);
    }

    @Test void silentLeaderboardDoesNothing() throws Exception {
        Leaderboard lb = new SilentLeaderboard();
        lb.save(entry());
        assertThat(lb.entries()).isEmpty();
    }
}

File: src/test/kotlin/com/jacuum/leaderboard/JsonLeaderboardTest.kt
package com.jacuum.leaderboard

import com.jacuum.algo.Direction
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path
import java.time.Instant

class JsonLeaderboardTest {

    @TempDir
    lateinit var tmp: Path

    private fun entry() = LeaderboardEntry(
        "id1", "Luke", "\uD83E\uDD16", "hash1", "SMALL", "Random",
        42, 100, 4200, Instant.now().toString(),
        listOf(TraceEvent(1, Direction.EAST, 1, 1, 100))
    )

    @Test fun saveAndRetrieveEntry() {
        val file = tmp.resolve("lb.json")
        val lb: Leaderboard = JsonLeaderboard(file)
        lb.save(entry())
        val entries = lb.entries()
        assertThat(entries).hasSize(1)
        assertThat(entries[0].username).isEqualTo("Luke")
    }

    @Test fun persistsAcrossInstances() {
        val file = tmp.resolve("lb2.json")
        JsonLeaderboard(file).save(entry())
        val entries = JsonLeaderboard(file).entries()
        assertThat(entries).hasSize(1)
    }

    @Test fun silentLeaderboardDoesNothing() {
        val lb: Leaderboard = SilentLeaderboard()
        lb.save(entry())
        assertThat(lb.entries()).isEmpty()
    }
}
